rllib/env/base_env.py

```diff
@@ -586,3 +621,124 @@ def reset(self) -> MultiAgentDict:
         self.last_dones = {"__all__": False}
         self.last_infos = {}
         return self.last_obs
+
+
+def convert_to_base_env(
+        env: EnvType,
+        make_env: Callable[[int], EnvType] = None,
+        num_envs: int = 1,
+        remote_envs: bool = False,
+        remote_env_batch_wait_ms: int = 0,
+        policy_config: Optional[PartialTrainerConfigDict] = None,
+) -> "BaseEnv":
+    """Converts an RLlib-supported env into a BaseEnv object.
+
+    Supported types for the `env` arg are gym.Env, BaseEnv,
+    VectorEnv, MultiAgentEnv, ExternalEnv, or ExternalMultiAgentEnv.
+
+    The resulting BaseEnv is always vectorized (contains n
+    sub-environments) to support batched forward passes, where n may also
+    be 1. BaseEnv also supports async execution via the `poll` and
+    `send_actions` methods and thus supports external simulators.
+
+    TODO: Support gym3 environments, which are already vectorized.
+
+    Args:
+        env: An already existing environment of any supported env type
+            to convert/wrap into a BaseEnv. Supported types are gym.Env,
+            BaseEnv, VectorEnv, MultiAgentEnv, ExternalEnv, and
+            ExternalMultiAgentEnv.
+        make_env: A callable taking an int as input (which indicates the
+            number of individual sub-environments within the final
+            vectorized BaseEnv) and returning one individual
+            sub-environment.
+        num_envs: The number of sub-environments to create in the
+            resulting (vectorized) BaseEnv. The already existing `env`
+            will be one of the `num_envs`.
+        remote_envs: Whether each sub-env should be a @ray.remote actor.
+            You can set this behavior in your config via the
+            `remote_worker_envs=True` option.
+        remote_env_batch_wait_ms: The wait time (in ms) to poll remote
+            sub-environments for, if applicable. Only used if
+            `remote_envs` is True.
+        policy_config: Optional policy config dict.
+
+    Returns:
+        The resulting BaseEnv object.
+    """
+
+    from ray.rllib.env.remote_vector_env import RemoteBaseEnv
+    from ray.rllib.env.external_env import ExternalEnv, ExternalEnvWrapper
+    from ray.rllib.env.multi_agent_env import MultiAgentEnv, \
+        MultiAgentEnvWrapper
+    from ray.rllib.env.vector_env import VectorEnv, VectorEnvWrapper
+    if remote_envs and num_envs == 1:
+        raise ValueError("Remote envs only make sense to use if num_envs > 1 "
+                         "(i.e. vectorization is enabled).")
+
+    # Given `env` is already a BaseEnv -> Return as is.
+    if isinstance(env, BaseEnv):
+        return env
+
+    # `env` is not a BaseEnv yet -> Need to convert/vectorize.
+
+    # MultiAgentEnv (which is a gym.Env).
+    if isinstance(env, MultiAgentEnv):
```

Comment:
I think things will be even cleaner if we can create some kind of common module level apis for these Envs.
Something like the following:

- external_env.py
    - def is_env_type(env) -> bool: return isinstance(env, ExternalEnv)
    - def to_base_env(env, make_env, num_envs, ...) -> BaseEnv:
           ... logics for converting ExternalEnv to BaseEnv.
- multi_agent_env.py
    - def is_env_type(env) -> bool: return isinstance(env, MultiAgentEnv)
    - def to_base_env(env, make_env, num_envs, ...) -> BaseEnv:
           ... logics for converting MultiAgentEnv to BaseEnv.
- vector_env.py
    - def is_env_type(env) -> bool: return isinstance(env, VectorEnv)
    - def to_base_env(env, make_env, num_envs, ...) -> BaseEnv:
           ... logics for converting VectorEnv to BaseEnv.
- generic_env.py
    - def is_env_type(env) -> bool: return isinstance(env, EnvType)
    - def to_base_env(env, make_env, num_envs, ...) -> BaseEnv:
           ... logics for converting basically gym.Env to BaseEnv.
           ... basically the logics in the last big else clause.

Then with this structure, we can clean up this function to be as simple as:

```
def convert_to_base_env(env, ...) -> BaseEnv:
    if isinstance(env, BaseEnv):
        return env

    from ray.rllib.env import external_env, multi_agent_env, vector_env, generic_env
    for env_type in [external_env, multi_agent_env, vector_env, generic_env]:
        if env_type.is_env_type(env):
            return env_type.to_base_env(env, make_env, ...)

    raise ValueError("Unknown Env type: ", type(env))
```

Let me know if there is anything that's not clear here.
I can also help send a patch if it makes things faster.

Reply:
Not sure about the `is_env_type()` methods. Wouldn't these just add extra weight to the code?
One could also just do (w/o needing these methods):
```
def convert_to_base_env(env, ...) -> BaseEnv:
    if isinstance(env, BaseEnv):
        return env

    from ray.rllib.env import external_env, multi_agent_env, vector_env, generic_env
    for env_type in [external_env, multi_agent_env, vector_env, generic_env]:
        if isinstance(env, env_type):
            return env_type.to_base_env(env, make_env, ...)

    raise ValueError("Unknown Env type: ", type(env))
```

Reply:
Happy to move the `to_base_env` logics into the individual classes. That makes a lot of sense.
Every env-class that has such a method is then directly visible as RLlib-supported.

Reply:
@avnishn , we can do this in a follow-up PR, though. This one already has lots of changes in it.

Reply:
is there a reason why we can't instead keep make this a static but standalone function, like we do in this PR anyways? The reason that we have this to_base_env function right now is because RLlib requires a BaseEnv for most of its operations. Is that reason enough that we should have this to_base_env function?

Perhaps it is cleaner, but it does definitely add weight to the class overall.

Reply:
I like the above solution though, because it removes the need for circular imports in the BaseEnv class

Reply:
we could probably call it something else? to_base_env implies that the function doesn't return a new class. Maybe we could call it `get_base_env` or `get_new_base_env`

Reply:
good point actually. I think if we can't delete the static BaseEnv.to_base_env() anyways, we might as well just keep using it. No need to add the top-level convert_to_base_env().
the part I care about is to keep BaseEnv.to_base_env() as minimal as possible, and make per-Env conversion logic live with those specific modules.

regarding is_env_type(), that's just so we don't have to import 4 Env classes using 4 import statements. and it provides a little bit of flexibility when you need more logics to decide whether a specific Env should handle the conversion. I won't insist on this either. the code will look like:

```
def convert_to_base_env(env, ...) -> BaseEnv:
    if isinstance(env, BaseEnv): return env

    from ray.rllib.env.external_env import ExternalEnv
    from ray.rllib.env.multi_agent_env import MultiAgentEnv
    ... import VectorEnv
    ... import generic_env_to_base_env

    for env_type in [ExternalEnv, MultiAgentEnv, VectorEnv]:
        if isinstance(env, env_type):
            return env_type.to_base_env(env, make_env, ...)

     return generic_env_to_base_env(env, make_env, ...)
```
Just a little less concise.

Either way is fine. Let's worry about it with the followup PR.